Add unit tests for Usage Tracking class

alexsanford opened the pull request:
Primarily testing generic functionality such as:

- cron job setup and execution
- ajax handler for enabling/disabling tracking
- sending the tracking request to Tracks when tracking is enabled
- displaying tracking opt-in dialog when appropriate

As this is my first significant test suite written for a plugin, I would appreciate any comments/suggestions regarding best practices.

Please ensure the tests pass in your dev environment. They pass on my two VM's (PHP 7.1 and 5.3) and on Circle CI (PHP 5.4). My VM's and CI are running against the latest version of WordPress, AFAICT.

donnapep commented on tests/unit-tests/test-class-usage-tracking.php:
```diff
@@ -0,0 +1,318 @@
+class Sensei_Usage_Tracking_Test extends WP_UnitTestCase {
+	public function setup() {
+		parent::setup();
+		$this->factory = new Sensei_Factory();
```
It doesn't look like you're testing course or lessons data, so you don't need to include `Sensei_Factory`.

alexsanford replied:
Removed

donnapep commented on tests/unit-tests/test-class-usage-tracking.php:
```diff
@@ -0,0 +1,318 @@
+	/**
+	 * Ensure cron job action is set up.
+	 */
+	public function test_cron_job_action_added() {
```
You tend to see camel case being used for PHPUnit function names. Since it's the style that the rest of Sensei's tests use as well, probably better to use camel case.

Also, you can use the `@covers` annotation to show which functions are being tested as part of this test case, since it's unclear from the function name and I'd need to look inside the function to figure it out. I usually try to do both (`@covers` + function name).

BTW, this test is going to need to be affected by https://github.com/Automattic/sensei/pull/1983. 😬

alexsanford replied:
Switched to camel case (I was debating this, and didn't notice that the rest of Sensei used camel case for tests. Good catch!)

Added `@covers` annotation.

Re: #1983 now that it's merged, I'll rebase this and update accordingly.

donnapep commented on tests/unit-tests/test-class-usage-tracking.php:
```diff
@@ -0,0 +1,318 @@
+		} );
+
+		// Should successfully schedule the task
+		$this->assertFalse( wp_get_schedule( 'sensei_core_jobs_usage_tracking_send_data' ) );
```
When doing multiple assertions in a test, it's helpful to add a comment as the last parameter that indicates what the assertion is testing (see https://github.com/Automattic/sensei/pull/1983/files#diff-b1c3ae0ceedad75266d017a21c49a56fR28). Otherwise, if an assertion in a test case fails, you'll need to spend time pinpointing which one it was.

alexsanford replied:
Good call. Messages added.

donnapep commented on tests/unit-tests/test-class-usage-tracking.php:
```diff
@@ -0,0 +1,318 @@
+		$this->assertFalse( wp_get_schedule( 'sensei_core_jobs_usage_tracking_send_data' ) );
+		$this->usage_tracking->maybe_schedule_tracking_task();
+		$this->assertNotFalse( wp_get_schedule( 'sensei_core_jobs_usage_tracking_send_data' ) );
+		$this->assertEquals( 1, $event_count );
```
An alternative to this may be to use `wp_next_scheduled` and then just compare return values between calls to ensure the timestamp is the same. It would eliminate the need for this bit - https://github.com/Automattic/sensei/pull/1984/files#diff-0ea0511e3f117834481aeca98312467fR31.

alexsanford replied:
I'd love to eliminate the filter. The problem with using `wp_next_scheduled`, however, is that in this case the "next scheduled" job is always going to be the first one that was added, regardless of whether more were scheduled after it. So I can't use it to tell whether additional jobs were scheduled, as far as I know.

donnapep replied:
Oh, that's too bad. 🙁

donnapep commented on tests/unit-tests/test-class-usage-tracking.php:
```diff
@@ -0,0 +1,318 @@
+	/**
+	 * Helpers for ajax request.
+	 */
+	private function setup_ajax_request() {
```
I'm not sure I understand what part of the usage tracking code this is testing since it doesn't appear to be calling any usage tracking functions?

alexsanford replied:
It's a helper method, not a test method. I moved it to the bottom of the file to make it more clear (as opposed to being at the end of the section where it is used)

donnapep replied:
Ah, right. Thx.

donnapep commented on tests/unit-tests/test-class-usage-tracking.php:
```diff
@@ -0,0 +1,318 @@
+		Sensei()->settings->set( 'sensei_usage_tracking_enabled', true );
+		Sensei()->settings->get_settings();
+
+		$this->expectOutputString( '' );
```
I'm not quite following these last few tests and why `expectOutputString` is being called before the call to `maybe_display_tracking_opt_in`. Could you clarify?

alexsanford replied:
It's because of the way [output testing](https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.output) works in phpunit. You need to set up the expectation first (probably because it starts the buffering at this point) and then call the code that produces the output.

donnapep replied:
Oh, interesting! TIL.

donnapep commented on tests/unit-tests/test-class-usage-tracking.php:
```diff
@@ -1,28 +1,31 @@
 
-	public function teardown() {
-		parent::teardown();
+	public function tearDown() {
```
If you're not doing anything in the `tearDown` function, then you don't need to include it. `parent::teardDown` gets called implicitly.

alexsanford commented:
Rebased onto `master` and fixed a few issues. Also fixed some PHP <5.4 compatibility issues in another test.

@donnapep could you please give this another sanity check with the new changes before I merge? I did a force push, so you'll have to reset your branch.

donnapep commented:
@alexsanford Yup, looks good! 👍
